File: rustnlpjava/src/main/java/dev/gigapixel/rustnlp/Strings.java
package dev.gigapixel.rustnlp;

import dev.gigapixel.rustnlp.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class Strings {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    

    static class StringsCleaner implements Runnable {

        MemorySegment segment;
        StringsCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            rustnlp_h.Strings_destroy(this.segment);
        }
    }

    Strings() {}
    Strings(MemorySegment handle, List<Object> selfEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        

    }
    
    
}

File: rustnlpkotlin/src/main/kotlin/dev/gigapixel/rustnlp/Strings.kt
package dev.gigapixel.rustnlp;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure


internal interface StringsLib: Library {
    fun Strings_destroy(handle: Pointer)
}

class Strings internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>
)  {

    internal class StringsCleaner(val handle: Pointer, val lib: StringsLib) : Runnable {
        override fun run() {
            lib.Strings_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<StringsLib> = StringsLib::class.java
        internal val lib: StringsLib = Native.load("rustnlp", libClass)
    }

}
